src/server/pps/server/api_server.go

```diff
@@ -3012,12 +3012,21 @@ func (a *apiServer) listPipeline(ctx context.Context, request *pps.ListPipelineR
 			})
 		})
 	}
+	var offset int64
+	if request.Page != nil {
+		offset = request.Page.PageIndex * request.Page.PageSize
+	}
+	var seen int64
+	var emmitted int64
 	if err := ppsutil.ListPipelineInfo(ctx, a.pipelines, request.Pipeline, request.History, func(pi *pps.PipelineInfo) error {
 		if ok, err := filterPipeline(ctx, pi); err != nil {
 			return errors.Wrap(err, "error filtering pipeline")
 		} else if !ok {
 			return nil
 		}
+		if request.Page != nil && request.Page.PageSize > 0 && emmitted >= request.Page.PageSize {
```

Comment:
This could just be `if pageSize := request.GetPage().GetPageSize(); pageSize > 0 && emitted >= pageSize {`.